blockchain-updates/src/main/scala/com/wavesplatform/events/BlockchainUpdates.scala

```diff
@@ -0,0 +1,157 @@
+    import scala.collection.JavaConverters._
+
+    val props = new Properties()
+    props.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, settings.bootstrapServers)
+    props.put(ConsumerConfig.GROUP_ID_CONFIG, "admin")
+    props.put(ConsumerConfig.CLIENT_ID_CONFIG, settings.clientId)
+    props.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "false")
+    props.put(ConsumerConfig.SESSION_TIMEOUT_MS_CONFIG, "10000")
+    props.put(ConsumerConfig.MAX_POLL_RECORDS_CONFIG, "1")
+
+    // read height and whether last event in Kafka is MicroBlock or MicroBlockRollback
+    val consumer = new KafkaConsumer[Unit, Int](
+      props,
+      new Deserializer[Unit] {
+        override def configure(configs: util.Map[String, _], isKey: Boolean): Unit = {}
+        override def deserialize(topic: String, data: Array[Byte]): Unit           = {}
+        override def close(): Unit                                                 = {}
+      },
+      new Deserializer[Int] {
+        override def configure(configs: util.Map[String, _], isKey: Boolean): Unit = {}
+        override def deserialize(topic: String, data: Array[Byte]): Int =
+          PBBlockchainUpdated.parseFrom(data).height
+        override def close(): Unit = {}
+      }
+    )
+
+    try {
+      consumer.partitionsFor(settings.topic).asScala.headOption.flatMap { partition =>
+        val tp = new TopicPartition(settings.topic, partition.partition)
+        consumer.assign(util.Arrays.asList(tp))
+        consumer.seek(tp, consumer.endOffsets(util.Arrays.asList(tp)).asScala.apply(tp) - 1)
+
+        val records = consumer.poll(JDuration.ofMillis(timeout.toMillis))
+
+        if (records.isEmpty) None
+        else records.records(tp).asScala.lastOption.map(_.value)
+      }
+    } catch { case _: Throwable => None }
+  }
+
+  @throws[IllegalStateException]("if events in Kafka are different from node state and it's impossible to recover")
+  private[this] def startupCheck(): Unit = {
+    // if kafkaHeight <= blockchainHeight — rollback node with event sending to Kafka. If rollback fails — fail
+    // if kafkaHeight > blockchainHeight — fail. This should not happen
+
+    // For later (maybe):, blockchain consistency can be checked better using Kafka view of blocks with signatures
+
+    val blockchainHeight = context.blockchain.height
+    getLastHeight() match {
+      case Some(kafkaHeight) =>
+        if (kafkaHeight <= (blockchainHeight + 1)) {
+          /*
+          Always rollback node on startup at least 1 block, since node loses ngState
+          and it's hard to tell whether the block in Kafka is full or not.
+
+          Rollback node to (kafkaHeight - 1) to be sure, since previous block is guaranteed to be solid.
+           */
+
+          val heightToRollbackTo = Math.max(kafkaHeight - 1, 1)
+
+          context.blockchain
+            .blockHeaderAndSize(heightToRollbackTo)
+            .toRight(new IllegalStateException(s"No block at height $heightToRollbackTo"))
```

Comment:
Use `Try[A]` instead of `Either[Throwable, A]`

Reply:
`blockchain.removeAfter(sigToRollback)` method returns Either. I used Either here to make it flatMap-compatible.